### PR TITLE
translate api/ to ts

### PR DESCRIPTION
**Pull request by stas-vilchik**

### CODE REVIEW

**Comment by stas-vilchik** [comment]
@gregaubert and one more "small" PR for you 😉

File: server/sonar-web/src/main/js/api/settings.ts [hunk 1]
```diff
@@ -18,38 +18,34 @@
...
   const { key } = definition;
-  const data = { key };
+  const data: RequestData = { key };
 
   if (definition.multiValues) {
     data.values = value;
   } else if (definition.type === TYPE_PROPERTY_SET) {
     data.fieldValues = value
-      .map(fields => omitBy(fields, value => value == null))
+      .map((fields: any) => omitBy(fields, value => value == null))
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/api/settings.ts]: CRITICAL Shadowed variable: 'value' [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1524)

File: server/sonar-web/src/main/js/api/issues.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  return getFacet(query, 'severities').then(r => r.facet);
+}
+
+export function getTags(query: RequestData): Promise<any> {
+  return getFacet(query, 'tags').then(r => r.facet);
+}
+
+export function extractAssignees(facet: Array<{ val: string }>, response: IssuesResponse): any {
+  return facet.map(item => {
+    const user = response.users ? response.users.find(user => user.login === item.val) : null;
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/api/issues.ts]: CRITICAL Shadowed variable: 'user' [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1524)

File: server/sonar-web/src/main/js/api/rules.ts [hunk 3]
```diff
@@ -17,14 +17,13 @@
...
-  const url = '/api/rules/search';
-  return getJSON(url, data);
+export function searchRules(data: RequestData) {
+  return getJSON('/api/rules/search', data);
 }
 
-export function takeFacet(response, property) {
-  const facet = response.facets.find(facet => facet.property === property);
+export function takeFacet(response: any, property: string) {
+  const facet = response.facets.find((facet: any) => facet.property === property);
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/api/rules.ts]: CRITICAL Shadowed variable: 'facet' [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1524)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 3 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: server/sonar-web/src/main/js/api/projectActivity.ts [hunk 4]
```diff
@@ -17,53 +17,42 @@
...
 
-export function getProjectActivity(
-  data /*: GetProjectActivityOptions */
-) /*: Promise<GetProjectActivityResponse> */ {
+export function getProjectActivity(data: {
+  project: string;
+  category?: string;
+  p?: number;
+  ps?: number;
+}): Promise<IGetProjectActivityResponse> {
```
Comment (gregaubert) [inline server/sonar-web/src/main/js/api/projectActivity.ts]: Missing the `| Response` type for when it throws

File: server/sonar-web/src/main/js/helpers/request.ts [hunk 5]
```diff
@@ -44,59 +36,53 @@ export function getCSRFTokenValue() /*: string */ {
...
-export function getCSRFToken() /*: Object */ {
+export function getCSRFToken(): { [x: string]: string } {
   // Fetch API in Edge doesn't work with empty header,
   // so we ensure non-empty value
   const value = getCSRFTokenValue();
   return value ? { [getCSRFTokenName()]: value } : {};
 }
 
-export function omitNil(obj /*: Object */) /*: Object */ {
+export function omitNil(obj: { [x: string]: any }): { [x: string]: any } {
```
Comment (gregaubert) [inline server/sonar-web/src/main/js/helpers/request.ts]: `RequestData` can be used here

File: server/sonar-web/src/main/js/api/projectActivity.ts (hunk 4 above)
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/api/projectActivity.ts]: So far TS is unable to track types for fulfilled and rejected promises separately. So there is no much sense to add `| Response` here, it won't give better types. I've added it only where TS was complaining.

Comment (gregaubert) [inline server/sonar-web/src/main/js/api/projectActivity.ts]: Ok good to know!

**Pull request merged**